remotes/docker/fetcher.go

```diff
@@ -206,3 +197,23 @@ func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string,
 
 	return resp.Body, nil
 }
+
+func getResponseErr(ctx context.Context, req *request, resp *http.Response) error {
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusNotFound {
+		return errors.Wrapf(errdefs.ErrNotFound, "content at %v not found", req.String())
+	}
+	var registryErr Errors
+	dec := json.NewDecoder(resp.Body)
+	if err := dec.Decode(&registryErr); err != nil || registryErr.Len() < 1 {
+		// In some cases, the error may just be a simple message rather than a json encoded string.
+		sb := &strings.Builder{}
+		_, err2 := io.Copy(sb, io.LimitReader(io.MultiReader(dec.Buffered(), resp.Body), 16*1024))
+		if err2 != nil {
+			log.G(ctx).WithError(err2).Debug("Error copying response body details")
+		}
+		return errors.Errorf("unexpected status code %v: %v, details: %s", req.String(), resp.Status, sb)
```

Comment:
I would rather us type this error, shoveling the raw http body into an error string still doesn't seem right. The cases this occurs is if the registry is not implementing the same error handling format or the response is from middleware. The output is more likely to be mangled html than a valid error message.

We could wrap the registry error type in a response error which holds the request, status code, registry error type. In this case, we can use the unknown registry error type and let the caller decide if it wants to present details.

Reply:
> The output is more likely to be mangled html than a valid error message.

Distribution spec is a bit sketchy on that ("If the response body is in JSON format"); i.e. it mentions the body to be JSON format, but doesn't mention wether or not that requires the content-type to be `application/json` (i.e., should a `text/plain` response be attempted to parse as JSON?)

> We could wrap the registry error type in a response error which holds the request, status code, registry error type. In this case, we can use the unknown registry error type and let the caller decide if it wants to present details.

You mean; preserve the body, but convert it to a "valid" format? I think that would work; at least the information would not be _lost_, and clients should be able to handle the standard format (how to handle other formats is somewhat "unspecified":  _"A 4XX response code from the registry MAY return a body in any format"_ )

Reply:
@tonistiigi additional thoughts on the above? ^^

Reply:
@dmcgowan That's a fair point. It also kinda sucks to expect all callers to be aware of this to make that determination.
I'd be fine dropping that, or possibly checking explicitly `text/plain` as the content type.

Reply:
Can you just dump the body into the log and mention the status code is unexpected and the error is not formatted.